index.d.ts

```diff
@@ -1,288 +1,523 @@
 
-  export type SortingRules<D> = SortingRule<D>[]
+  export type Hooks<D extends object = {}> = UseTableHooks<D> //&
+  // UseSortByHooks<D> &
+  // UseGroupByHooks<D> &
+  // UseExpandedHooks<D> &
+  // UseRowSelectHooks<D>
 
-  export type SortByFn = (a: any, b: any, desc: boolean) => 0 | 1 | -1
+  export type Cell<D extends object = {}> = UseTableCellProps<D>
 
-  export type Filters<D> = Record<IdType<D>, string>
+  export type Column<D extends object = {}> = UseTableColumnOptions<D>
 
-  export interface Cell<D = {}> extends TableInstance<D> {
-    cell: { value: any }
-    column: Column<D>
-    row: Row<D>
-    render: (type: 'Cell' | 'Aggregated', userProps?: any) => any
-    isGrouped?: boolean
-    isAggregated?: boolean
-    isRepeatedValue?: boolean
-    getCellProps: () => any
+  export type ColumnInstance<D extends object = {}> = Column<D> &
+    UseTableColumnProps<D>
+
+  export type HeaderGroup<D extends object = {}> = UseTableHeaderGroupProps<D>
+
+  export type Row<D extends object = {}> = UseTableRowProps<D>
+
+  /* #region useTable */
+  export function useTable<D extends object = {}>(
+    options: TableOptions<D>,
+    ...plugins: PluginHook<D>[]
+  ): TableInstance<D>
+
+  export type UseTableOptions<D extends object> = Partial<{
+    // columns: Column<D>[] // FIXME
+    // state: TableStateTuple<D> // FIXME
+    data: D[]
+    defaultColumn: Partial<Column<D>>
+    initialRowStateKey: IdType<D>
+    getSubRows: (row: Row<D>, relativeIndex: number) => Row<D>[]
+    getRowID: (row: Row<D>, relativeIndex: number) => string
+    debug: boolean
+  }>
+
+  export interface UseTableHooks<D extends object> {
+    columnsBeforeHeaderGroups: ((
+      flatColumns: Column<D>[],
+      instance: TableInstance<D>
+    ) => Column<D>[])[]
+    columnsBeforeHeaderGroupsDeps: ((
+      deps: any[],
+      instance: TableInstance<D>
+    ) => any[])[]
+    useMain: ((instance: TableInstance<D>) => TableInstance<D>)[]
+    useRows: ((rows: Row<D>[], instance: TableInstance<D>) => Row<D>[])[]
+    prepareRow: ((row: Row<D>, instance: TableInstance<D>) => Row<D>)[]
+
+    // Prop Hooks
+    getTableProps: ((instance: TableInstance<D>) => object)[]
+    getRowProps: ((row: Row<D>, instance: TableInstance<D>) => object)[]
+    getHeaderGroupProps: ((
+      headerGroup: HeaderGroup<D>,
+      instance: TableInstance<D>
+    ) => object)[]
+    getHeaderProps: ((
+      column: Column<D>,
+      instance: TableInstance<D>
+    ) => object)[]
+    getCellProps: ((cell: Cell<D>, instance: TableInstance<D>) => object)[]
   }
 
-  export interface Row<D = {}> {
-    index: number
+  export type UseTableColumnOptions<D extends object> = Accessor<D> &
+    Partial<{
+      columns: Column<D>
+      show: boolean | ((instance: TableInstance<D>) => boolean)
+      Header:
+        | ComponentType<TableInstance<D> & { column: Column<D> }>
+        | ReactNode
+      Cell: ComponentType<TableInstance<D> & { cell: Cell<D> }> | ReactNode
```

Comment:
The Cell one is weird, columns and row gets repeated everywhere.

```tsx
Cell: ComponentType<TableInstance<D> & { cell: Cell<D>, column: Column<D>, row: Row<D> }> | ReactNode
```

Reply:
maybe extract `TableInstance<D> & { column: Column<D> }`, it gets passed to every renderer except the cell one, and also gets passed through to the filter calls (UseFiltersColumnOptions.Filter for example).

Reply:
It would be nice to refactor these a bit so that implementation of cell or header renderers can have a tight props definition, maybe HeaderProps and CellProps, or leave the external definition as Header and Cell, but call the internal version something else since the internal ones are generally needed to cleanly define chunks of instance, and aren't very likely to need to be typed by the user.